pkg/ebpftracer/signature/socks5_detected.go

```diff
@@ -0,0 +1,171 @@
+package signature
+
+import (
+	"fmt"
+
+	v1 "github.com/castai/kvisor/api/v1/runtime"
+	"github.com/castai/kvisor/pkg/ebpftracer/events"
+	"github.com/castai/kvisor/pkg/ebpftracer/types"
+	"github.com/castai/kvisor/pkg/net/packet"
+	"github.com/castai/kvisor/pkg/proc"
+	"github.com/elastic/go-freelru"
+)
+
+var _ Signature = (*SOCKS5Detected)(nil)
+
+type SOCKS5DetectionState uint8
+
+const (
+	SOCKS5Unknown SOCKS5DetectionState = iota
+	SOCKS5InitialClientRequestReceived
+	SOCKS5InitialClientRequestSend
+	SOCKS5InitialServerResponseSend
+	SOCKS5InitialServerResponseReceived
+)
+
+const DefaultSOCKS5SignatureCacheSize = 1024
+
+type SOCKS5DetectionSignatureConfig struct {
+	CacheSize uint32
+}
+
+type SOCKS5Detected struct {
+	detectionStateCache freelru.Cache[proc.PID, SOCKS5DetectionState]
+}
+
+func NewSOCKS5DetectedSignature(cfg SOCKS5DetectionSignatureConfig) (Signature, error) {
+	var cacheSize uint32
+
+	if cacheSize <= 0 {
+		cacheSize = DefaultSOCKS5SignatureCacheSize
+	} else {
+		cacheSize = cfg.CacheSize
+	}
+
+	cache, err := freelru.NewSynced[proc.PID, SOCKS5DetectionState](cacheSize, func(u uint32) uint32 {
```

Comment:
Now this means there is only one socket5 socket per pid. Would be safer to include pid+fd to make sure it's unique.

Reply:
Good point! This logic would definitely fall over, when a single process is both SOCKS5 client/server. Two problems though: 
* We would need to rework how we report events from network related eBPF
* We do not have the FD easily available 

For now I would go with just the PID and if we see this is a limitation worth fixing, invest some more time 🙂 
What do you think?

Reply:
Sounds good